usr/lib/common/attributes.c

```diff
@@ -328,10 +328,30 @@ CK_BBOOL compare_attribute_array(CK_ATTRIBUTE_PTR a1, CK_ULONG a1_len,
 }
 
 #ifdef DEBUG
+void dump_array_attr(CK_ATTRIBUTE_PTR a)
+{
+    CK_ATTRIBUTE_PTR attrs;
+    CK_ULONG num_attrs, i;
+    const char *typestr = p11_get_cka(a->type);
+
+    attrs = (CK_ATTRIBUTE_PTR)a->pValue;
+    num_attrs = a->ulValueLen / sizeof(CK_ATTRIBUTE);
+
+    TRACE_DEBUG("  %s: begin array: num_elements=%lu\n", typestr, num_attrs);
+    for (i = 0; i < num_attrs; i++)
+        dump_attr(&attrs[i]);
+    TRACE_DEBUG("  %s: end array\n", typestr);
+}
+
 void dump_attr(CK_ATTRIBUTE_PTR a)
 {
     const char *typestr = p11_get_cka(a->type);
```

Comment:
Since `p11_get_cka` is not pure, the compiler has to at least call it (maybe not initialize `typestr` and defer that, but I doubt that will give any benefit).  Maybe we should initialize `typestr` after dealing with array attributes?

Reply:
Moved initialization below the array attribute handling.